### PR TITLE
KAFKA-1595; Remove deprecated and slower Scala JSON parser

### PR DESCRIPTION
**Pull request by ijuma**
In a test by @onurkaraman involving 3066 topics and 95895 partitions,
Controller initialisation time spent on JSON parsing would be reduced from
37.1 seconds to 0.7 seconds by switching from the current JSON parser to
Jackson. See the following JIRA comment for more details:

https://issues.apache.org/jira/browse/KAFKA-5328?focusedCommentId=16027086

I tested that we only use Jackson methods introduced in 2.0 in the main
codebase by compiling it with the older version locally. We use a
constructor introduced in 2.4 in one test, but I didn't remove it as it
seemed harmless. The reasoning for this is explained in the mailing list
thread:

http://search-hadoop.com/m/uyzND1FWbWw1qUbWe

Finally, this PR only handles the parsing side. It would be good to use Jackson
for serialising to JSON as well. I filed KAFKA-5631 for that.

### CODE REVIEW

**Comment by ijuma** [comment]
Jenkins, retest this please

**Comment by guozhangwang** [comment]
@ijuma Is this still relevant? If yes could you rebase?

**Comment by ijuma** [comment]
@guozhangwang It is still relevant, but hopelessly out of date by now. :) I will close the PR until I have a chance to resubmit.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6037/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6022/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6038/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6023/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by ijuma** [comment]
@onurkaraman I've rebased the branch. Would you like to review it?

File: core/src/main/scala/kafka/server/DynamicConfigManager.scala [hunk 1]
```diff
@@ -89,39 +90,33 @@ class DynamicConfigManager(private val zkUtils: ZkUtils,
...
-            case _ => throw new IllegalArgumentException("Config change notification has an unsupported version " + map("version") +
-                "Supported versions are 1 and 2.")
-          }
-
-        case _ => throw new IllegalArgumentException("Config change notification has an unexpected value. The format is:" +
-          "{\"version\" : 1, \"entity_type\":\"topics/clients\", \"entity_name\" : \"topic_name/client_id\"}." + " or " +
-          "{\"version\" : 2, \"entity_path\":\"entity_type/entity_name\"}." +
-          " Received: " + json)
+      // There are no config overrides.
+      // Ignore non-json notifications because they can be from the deprecated TopicConfigManager
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/DynamicConfigManager.scala]: Will clarify this comment as the new placement makes it less clear.

**Comment by ijuma** [comment]
@hachikuji, if you have time, maybe you can take a look to see what you think of the approach?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6301/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6285/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6306/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6290/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by onurkaraman** [comment]
My bad @ijuma. I thought I had left comments shortly after you pinged me on the PR.

Sort of a high-level comment: It seems like the approach of using implicits would work. I think people have mixed opinions on implicits in scala. Have you considered a more explicit approach? For instance, using jackson to convert to-and-from versioned POJOs. I think this would end up having clearly defined data formats for each znode path similar to our Protocol class for wire format definitions.

**Comment by ijuma** [comment]
Thanks for the review @onurkaraman. There are many types of implicit usage in Scala. In this case I'm using implicits as type classes. This is generally considered as one of the most useful features in Scala (and Haskell). Implicit conversions are the ones that can be misused at times (they are still useful as extension methods).

I did consider versioned POJOs and I agree that it would be nice to have clearly defined formats. That's a bigger change, however and there's a higher potential for breaking compatibility with older formats depending on the quality of our tests (I'm a bit suspicious). We'd have to depend on the jackson-scala module if we use Scala classes in the POJOs, but we could potentially stick to Java classes.

 My thinking is that we should first convert to Jackson while using the same approach as we do now. It's easy to review and gives us the performance, concurrency and additional type safety benefit.

In a second phase, we should look into using versioned classes to represent the data types. At that point, we can also use Jackson to serialize to JSON. We can also do this incrementally as the underlying library is the same, we are just changing the implementation.

Thoughts?

**Comment by onurkaraman** [comment]
@ijuma: Makes sense. I agree.

**Comment by onurkaraman** [comment]
LGTM.

**Comment by ijuma** [comment]
@rajinisivaram, maybe you can review this? It's mostly mechanical changes with some thin wrappers over Jackson.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6348/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6333/
Test PASSed (JDK 8 and Scala 2.12).

File: core/src/main/scala/kafka/utils/Json.scala [hunk 2]
```diff
@@ -14,34 +14,27 @@
...
-      } catch {
-        case t: Throwable =>
-          throw new KafkaException("Can't parse json string: %s".format(input), t)
-      }
-    }
-  }
-  
+  def parseFull(input: String): Option[JsonValue] =
+    try Option(mapper.readTree(input)).map(JsonValue(_))
+    catch { case e: JsonProcessingException => None }
```
Comment (rajinisivaram) [inline core/src/main/scala/kafka/utils/Json.scala]: Isn't this swallowing an exception while the old code used to propagate parsing errors? Don't callers just ignore `None`? It looks like a deliberate change, but I wasn't sure why.

**Review by rajinisivaram — APPROVED** [review]
Left a minor comment, but the changes look good.

File: core/src/main/scala/kafka/utils/Json.scala (hunk 2 above)
Comment (ijuma) [reply core/src/main/scala/kafka/utils/Json.scala]: Thanks for the review @rajinisivaram. The behaviour is the same. The catch block in the old code doesn't seem to do anything. For example:

```scala
scala> JSON.parseFull("}")
res1: Option[Any] = None
```

Comment (ijuma) [reply core/src/main/scala/kafka/utils/Json.scala]: The example above shows that `None` is returned for invalid JSON.

Comment (rajinisivaram) [reply core/src/main/scala/kafka/utils/Json.scala]: @ijuma Thanks, LGTM.

**Comment by ijuma** [comment]
@rajinisivaram, I noticed a bug and fixed it here (with tests): https://github.com/apache/kafka/pull/83/commits/30c9d02e0b0c427a91843f37f9ff9c91826157cd

I also added scaladoc to the parsing classes and removed the unnecessary `JsonValue.toOption` method.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6384/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6369/
Test PASSed (JDK 8 and Scala 2.12).

**Review by rajinisivaram — APPROVED** [review]
LGTM

**Comment by ijuma** [comment]
Thanks for the reviews, merging to trunk.

**Comment by leonhong** [comment]
great pr

**Pull request closed without merging**